sdks/python/apache_beam/runners/interactive/caching/streaming_cache.py

```diff
@@ -0,0 +1,124 @@
+#
+# Licensed to the Apache Software Foundation (ASF) under one or more
+# contributor license agreements.  See the NOTICE file distributed with
+# this work for additional information regarding copyright ownership.
+# The ASF licenses this file to You under the Apache License, Version 2.0
+# (the "License"); you may not use this file except in compliance with
+# the License.  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+from __future__ import absolute_import
+
+from apache_beam.portability.api.beam_interactive_api_pb2 import InteractiveStreamHeader
+from apache_beam.portability.api.beam_interactive_api_pb2 import InteractiveStreamRecord
+from apache_beam.portability.api.beam_runner_api_pb2 import TestStreamPayload
+from apache_beam.utils import timestamp
+from apache_beam.utils.timestamp import Timestamp
+
+
+class StreamingCache(object):
+  """Abstraction that holds the logic for reading and writing to cache.
+  """
+  def __init__(self, readers):
+    self._readers = readers
+
+  class Reader(object):
+    """Abstraction that reads from PCollection readers.
+
+    This class is an Abstraction layer over multiple PCollection readers to be
+    used for supplying the Interactive Service with TestStream events.
+
+    This class is also responsible for holding the state of the clock, injecting
+    clock advancement events, and watermark advancement events.
+    """
+    def __init__(self, readers):
+      self._timestamp = Timestamp.of(0)
+      self._readers = {}
+      self._headers = {}
+      readers = [r.read() for r in readers]
+
+      # The header allows for metadata about an entire stream, so that the data
+      # isn't copied per record.
+      for r in readers:
+        header = InteractiveStreamHeader()
+        header.ParseFromString(next(r))
+
+        # Main PCollections in Beam have a tag as None. Deserializing a Proto
+        # with an empty tag becomes an empty string. Here we normalize to what
+        # Beam expects.
+        self._headers[header.tag if header.tag else None] = header
+        self._readers[header.tag if header.tag else None] = r
+
+      self._watermarks = {tag: timestamp.MIN_TIMESTAMP for tag in self._headers}
+
+    def read(self):
+      """Reads records from PCollection readers.
+      """
+      records = {}
+      for tag, r in self._readers.items():
```

Comment:
This logic is incorrect. One may have to return multiple events from one stream before any events of another. This should be tested.

Reply:
Good catch, I changed it to be able to read from multiple underlying caches. PTAL.

Reply:
I'm still having  hard time convincing myself that the logic between _read_next() and read() is correct. 

What you're really trying to do is perform a merge sort of the reader on timestamp. For readability and maintainability, it would probably be best expressed that way. Whether these are Events that already have the tag embedded (I'm assuming the backing file format will be compressed, so little to no overhead) or the input is a dict of {tag: iterator} and the tags applied when merging depends on which is easier (e.g. when writing).

Reply:
Ack, I broke it up to smaller methods, ptal.

Reply:
Thanks. The key to making things readable is not just having smaller methods, but having each piece a self-contained unit of abstraction. For example, making `target_timestamp` a parameter of _test_stream_events_before_target rather than setting a member on self and inspecting it in that function is a nice step forward. On the other hand, _min_timestamp_of now has an (implicit) assumption that the events are passed in (reversed) timestamp order which makes it harder to understand in isolation. 

I think, as far as I am able to follow it, the code as written is now correct.